lib/TfIdfVectorizer.jsm

```diff
@@ -0,0 +1,74 @@
+
+const REGEXP_SPLITS = new RegExp(`[${UNICODE_SPACE}${UNICODE_SYMBOL}${UNICODE_PUNCT}]+`);
+const REGEXP_ALPHANUMS = new RegExp(`^[${UNICODE_NUMBER}${UNICODE_MARK}${UNICODE_LETTER}]+$`);
+
+this.TfIdfVectorizer = class TfIdfVectorizer {
+  /**
+   * Downcases the text, and splits it into consecutive alphanumeric characters.
+   * This is locale aware, and so will not strip accents. This uses "word
+   * breaks", and os is not appropriate for languages without them
+   * (e.g. Chinese).
+   */
+  tokenize(text) {
+    return text.toLocaleLowerCase().split(REGEXP_SPLITS).filter(tok => tok.match(REGEXP_ALPHANUMS));
+  }
+
+  /**
+   * Converts a sequence of tokens into an L2 normed TF-IDF. Any terms that are
+   * not preindexed (i.e. does has a computed inverse document frequency) will
+   * be dropped.
+   */
+  toksTotfIdfVector(tokens, vocab_idfs) {
+    let tfidfs = {};
+
+    // calcualte the term frequencies
+    for (let i = 0; i < tokens.length; i++) {
+      let tok = tokens[i];
+      if (!(tok in vocab_idfs)) {
+        continue;
+      }
+      if (!(tok in tfidfs)) {
+        tfidfs[tok] = [vocab_idfs[tok][0], 1];
+      } else {
+        tfidfs[tok][1]++;
+      }
+    }
+
+    // now multiply by the log inverse document frequencies, then take
+    // the L2 norm of this.
+    let l2Norm = 0.0;
+    Object.keys(tfidfs).forEach(tok => {
+      tfidfs[tok][1] *= vocab_idfs[tok][1];
+      l2Norm += tfidfs[tok][1] * tfidfs[tok][1];
+    });
+    l2Norm = Math.sqrt(l2Norm);
+    Object.keys(tfidfs).forEach(tok => {
+      tfidfs[tok][1] /= l2Norm;
```

Comment:
Any chance to see `l2Norm === 0`? If so, better add a guard to prevent the divided-by-zero error.